### PR TITLE
python3Packages.packageurl-python: init at 0.9.4

### PR DESCRIPTION
**Pull request by armijnhemel**
###### Motivation for this change
packageurl (AKA purl) is an upcoming standard to describe packages in a more uniform way across operating systems and Linux distributions

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116325` at 4706bf95 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/bc18dbc9a846d3a26140929413d557ac">python38Packages.packageurl</a></li></li>
    <li><a href="https://gist.github.com/ca81d2a38efd37c394f677d3c9769b0f">python39Packages.packageurl</a></li></li>
  </ul>
</details>

File: pkgs/development/python-modules/packageurl-python/default.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+buildPythonPackage rec {
+  pname = "packageurl-python";
+  version = "0.9.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0mpvj8imsaqhrgfq1cxx16flc5201y78kqa7bh2i5zxsc29843mx";
+  };
+
+  pythonImportsCheck = [ "packageurl" ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/packageurl-python/default.nix:12]:
```suggestion
  pythonImportsCheck = [ "packageurl" ];
  
  checkInputs = [ pytestCheckHook ];
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/packageurl-python/default.nix (hunk 1 above)
Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: OK, added

**Comment by SuperSandro2000** [comment]
Please squash the review commit together as you can't do this from the github web ui without also squashing the maintainer add commit.

File: pkgs/development/python-modules/packageurl-python/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: You shoyld also try whether the tests run successfully.

Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: Loving it that people are giving conflicting information. Very helpful :-/

Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: > Loving it that people are giving conflicting information. Very helpful :-/

To be a bit more constructive: the previous time I submitted it I was told to include what I included if I was not able to run tests. I am not able to run the tests at this point.

Comment (dotlambda) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: Sorry, this was more about the missing argument `pytestCheckHook`.
If you post the output, we can help figure out how to make the tests run.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -4809,6 +4809,8 @@ in {
 
   oyaml = callPackage ../development/python-modules/oyaml { };
 
+  packageurl = callPackage ../development/python-modules/packageurl-python { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix:4812]:
```suggestion
  packageurl-python = callPackage ../development/python-modules/packageurl-python { };
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (armijnhemel) [reply pkgs/top-level/python-packages.nix:4812]: Sorry, but no. In a previous version of this PR I was asked to rename it from "packageurl-python" to "packageurl" ;-)

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:4812]: Didn't know that. I would say @rmcgibbo was wrong there, but maybe @SuperSandro2000, @FRidh, or @jonringer have an opinion on that. I thought the attribute name should always match `pname`.

Comment (armijnhemel) [reply pkgs/top-level/python-packages.nix:4812]: > Didn't know that. I would say @rmcgibbo was wrong there, but maybe @SuperSandro2000, @FRidh, or @jonringer have an opinion on that. I thought the attribute name should always match `pname`.

Personally I don't care what it is called but you should all make up your mind :-)

Comment (FRidh) [reply pkgs/top-level/python-packages.nix:4812]: The normalized PyPI name for both, so `packageurl-python`. This name is predictable and can be used by external tooling such as mach-nix.

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:4812]: clarifying this in https://github.com/NixOS/nixpkgs/pull/116344

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:4812]: Sorry for the back and forth @armijnhemel!

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 116325` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>python38Packages.packageurl</li>
<li>python39Packages.packageurl</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
python38Packages.packageurl:
<details>


warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
</details>

File: pkgs/development/python-modules/packageurl-python/default.nix [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{ buildPythonPackage, fetchPypi, lib }:
```
Comment (jonringer) [inline pkgs/development/python-modules/packageurl-python/default.nix:1]:
```suggestion
{ buildPythonPackage, fetchPypi, lib, pytestCheckHook }:
```

```
undefined variable 'pytestCheckHook' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/development/python-modules/packageurl-python/default.nix:12:19
```

Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix]: Urgh. that is stupid. Fixed.

File: pkgs/development/python-modules/packageurl-python/default.nix (hunk 1 above)
Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: I will need some help here. Note: I am not on NixOS, but Fedora. It has been 15 years since I did any serious work with Nix so my skills are very rusty.

Comment (dotlambda) [reply pkgs/development/python-modules/packageurl-python/default.nix:14]: I get
```
Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /build/packageurl-python-0.9.4, configfile: setup.cfg
collected 258 items

tests/test_packageurl.py .....................................           [ 14%]
tests/contrib/test_get_path_segments.py ...                              [ 15%]
tests/contrib/test_purl2url.py ..                                        [ 16%]
tests/contrib/test_url2purl.py ......................................... [ 32%]
........................................................................ [ 60%]
........................................................................ [ 87%]
...............................                                          [100%]

============================= 258 passed in 0.84s ==============================
Finished executing pytestCheckPhase
```
So everything seems to work.

**Review by dotlambda — CHANGES_REQUESTED** [review]
The only thing left to be done is squashing the last two commits into the second one and correct the commit message.

**Comment by armijnhemel** [comment]
grrr, sorry, I thought I had changed the commit message
